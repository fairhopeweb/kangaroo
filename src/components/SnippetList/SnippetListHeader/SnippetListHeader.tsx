import { memo, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Button, ButtonGroup, InputGroup } from '@blueprintjs/core';
import { GistsListResponseData } from '@octokit/types';

import { RootState, AppDispatch } from 'store/types';
import { showModal } from 'store/modal/actions';
import { setAuthToken, deleteAuthData } from 'store/auth/actions';
import { synchronizeGist, createBackupGist, addSnippet } from 'store/snippets/actions';
import { ACCOUNT_MODAL } from 'components/Modal/ModalOverlay/constants';

import styles from './SnippetListHeader.module.scss';

type SnippetListHeaderProps = {
  query: string;
  onSearchChange: (value: string) => void;
};

const SnippetListHeader: React.FC<SnippetListHeaderProps> = ({ query, onSearchChange }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { token } = useSelector((state: RootState) => state.auth);
  const [loading, setLoading] = useState(false);

  const handleAddSnippet = (): void => {
    dispatch(addSnippet());
  };

  const handleDeleteAuthData = (): void => {
    dispatch(deleteAuthData());
  };

  const handleSetAuthToken = (token: string): Promise<GistsListResponseData> => {
    return dispatch(setAuthToken(token));
  };

  const handleCreateBackupGist = (description: string, token: string): Promise<string> => {
    return dispatch(createBackupGist(description, token));
  };

  const handleSynchronizeGist = (backupLocalSnippets: boolean, token: string, id: string): Promise<string> => {
    return dispatch(synchronizeGist(backupLocalSnippets, token, id));
  };

  const handleSearchOnChange = ({ target: { value } }: React.ChangeEvent<HTMLInputElement>): void => {
    onSearchChange(value);
  };

  const handleClearSearch = (): void => {
    onSearchChange('');
  };

  const handleAccountModalOpen = (): void => {
    const dispatchShowModalAction = (): void => {
      dispatch(
        showModal(ACCOUNT_MODAL, {
          onSetAuthToken: handleSetAuthToken,
          onSynchronizeGist: handleSynchronizeGist,
          onCreateBackupGist: handleCreateBackupGist,
          onDeleteAuthData: handleDeleteAuthData,
        }),
      );
      setLoading(false);
    };

    setLoading(true);

    if (!token) {
      dispatchShowModalAction();
      return;
    }

    handleSetAuthToken(token).then(() => {
      dispatchShowModalAction();
    });
  };

  const renderClearSearchButton = (): React.ReactElement | undefined => {
    if (query) {
      return <Button icon="cross" minimal={true} onClick={handleClearSearch} />;
    }
  };

  return (
    <div className={styles.root}>
      <div className={styles.container}>
        <ButtonGroup minimal={true}>
          <Button icon="person" loading={loading} onClick={handleAccountModalOpen} />
          <Button icon="add-to-artifact" onClick={handleAddSnippet} />
        </ButtonGroup>
      </div>

      <div className={styles.searchContainer}>
        <InputGroup
          leftIcon="search"
          placeholder="Search..."
          rightElement={renderClearSearchButton()}
          fill={true}
          value={query}
          onChange={handleSearchOnChange}
        />
      </div>
    </div>
  );
};

export default memo(SnippetListHeader);
