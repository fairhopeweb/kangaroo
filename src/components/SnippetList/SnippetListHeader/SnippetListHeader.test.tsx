import { Provider } from 'react-redux';
import { configure, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configureMockStore } from '@jedmao/redux-mock-store';
import renderer from 'react-test-renderer';

import SnippetListHeader from './SnippetListHeader';

configure({ adapter: new Adapter() });

describe('<SnippetListHeader />', () => {
  const mockStore = configureMockStore();
  const mockProps = {
    query: '',
    onSearchChange: jest.fn(),
  };
  const initialState = { auth: { token: 'TOKEN' } };

  let store;

  it('renders without crashing', () => {
    store = mockStore(initialState);
    shallow(
      <Provider store={store}>
        <SnippetListHeader {...mockProps} />
      </Provider>,
    );
  });

  it('matches snapshot', () => {
    store = mockStore(initialState);
    const wrapper = renderer.create(
      <Provider store={store}>
        <SnippetListHeader {...mockProps} />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
