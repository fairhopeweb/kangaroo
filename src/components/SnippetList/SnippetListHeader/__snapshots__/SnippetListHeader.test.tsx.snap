// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SnippetListHeader /> matches snapshot 1`] = `
<div
  className="SnippetListHeader"
>
  <div
    className="SnippetListHeader--container"
  >
    <div
      className="bp3-button-group bp3-minimal"
    >
      <button
        className="bp3-button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="bp3-icon bp3-icon-person"
          icon="person"
        >
          <svg
            data-icon="person"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <path
              d="M15.68 14.32c-.46-1.05-2.68-1.75-4.16-2.4-1.48-.65-1.28-1.05-1.33-1.59-.01-.07-.01-.15-.01-.23.51-.45.92-1.07 1.19-1.78 0 0 .01-.04.02-.05.06-.15.11-.32.15-.48.34-.07.54-.44.61-.78.08-.14.23-.48.2-.87-.05-.5-.25-.73-.47-.82v-.09c0-.63-.06-1.55-.17-2.15A3.671 3.671 0 0010.32.72C9.68.25 8.79-.01 8-.01c-.79 0-1.68.25-2.31.73-.61.47-1.06 1.13-1.28 1.86-.05.17-.09.33-.11.5-.12.6-.17 1.51-.17 2.15v.08c-.24.09-.45.32-.5.83-.03.38.13.72.2.86.08.35.28.72.63.78.04.17.09.33.15.49 0 .01.01.02.01.03l.01.01c.27.72.7 1.35 1.22 1.8 0 .07-.01.14-.01.21-.05.54.1.94-1.37 1.59-1.48.65-3.7 1.35-4.16 2.4-.46 1.05-.27 1.67-.27 1.67h15.92c-.01.01.18-.61-.28-1.66z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
      <button
        className="bp3-button"
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="bp3-icon bp3-icon-add-to-artifact"
          icon="add-to-artifact"
        >
          <svg
            data-icon="add-to-artifact"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <path
              d="M14 4.01h-1v-1c0-.55-.45-1-1-1s-1 .45-1 1v1h-1c-.55 0-1 .45-1 1s.45 1 1 1h1v1c0 .55.45 1 1 1s1-.45 1-1v-1h1c.55 0 1-.45 1-1 0-.56-.45-1-1-1zm-13 2h6c.55 0 1-.45 1-1s-.45-1-1-1H1c-.55 0-1 .45-1 1s.45 1 1 1zm8 6H1c-.55 0-1 .45-1 1s.45 1 1 1h8c.55 0 1-.45 1-1 0-.56-.45-1-1-1zm0-4H1c-.55 0-1 .45-1 1s.45 1 1 1h8c.55 0 1-.45 1-1 0-.56-.45-1-1-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="SnippetListHeader--search-container"
  >
    <div
      className="bp3-input-group bp3-fill"
    >
      <span
        aria-hidden={true}
        className="bp3-icon bp3-icon-search"
        icon="search"
      >
        <svg
          data-icon="search"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M15.55 13.43l-2.67-2.68a6.94 6.94 0 001.11-3.76c0-3.87-3.13-7-7-7s-7 3.13-7 7 3.13 7 7 7c1.39 0 2.68-.42 3.76-1.11l2.68 2.67a1.498 1.498 0 102.12-2.12zm-8.56-1.44c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <input
        className="bp3-input"
        onChange={[Function]}
        placeholder="Search..."
        style={
          Object {
            "paddingLeft": undefined,
            "paddingRight": undefined,
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
